src/infra/core/src/repos/metadata_chain_visitor_impl.rs

```diff
@@ -0,0 +1,340 @@
+    next_data_event: Option<MetadataEvent>,
+}
+
+impl ValidateOffsetsAreSequentialVisitor {
+    fn process_first_block(&mut self, first_block: &MetadataBlock) -> Decision {
+        self.has_first_block_processed = true;
+
+        let Some(data_block) = first_block.as_data_stream_block() else {
+            // If it's not a data block, there's nothing to validate
+            return Decision::Stop;
+        };
+
+        self.next_prev_offset = data_block.event.prev_offset;
+        self.next_new_data = data_block.event.new_data.cloned();
+        self.next_data_event = Some(first_block.event.clone());
+
+        Decision::NextOfType(MetadataBlockTypeFlags::DATA_BLOCK)
+    }
+}
+
+impl MetadataChainVisitor for ValidateOffsetsAreSequentialVisitor {
+    fn visit(
+        &mut self,
+        (block, _): MetadataBlockWithOptionalHashRef,
+    ) -> Result<Decision, AppendError> {
+        // Only check AddData and ExecuteTransform.
+        // SetWatermark is also considered a data stream event but does not carry the
+        // offsets.
+        if !self.has_first_block_processed {
+            return Ok(self.process_first_block(block));
+        };
+
+        let Some(data_block) = block.as_data_stream_block() else {
+            return Ok(Decision::NextOfType(MetadataBlockTypeFlags::DATA_BLOCK));
+        };
+
+        // Validate input/output offset sequencing
+        let expected_next_prev_offset = data_block.event.last_offset();
+
+        if self.next_prev_offset != expected_next_prev_offset {
+            invalid_event!(
+                self.next_data_event.take().unwrap(),
+                "Carried prev offset does not correspond to the last offset in the chain",
+            );
+        }
+
+        // Validate internal offset consistency
+        if let Some(new_data) = &self.next_new_data {
+            let expected_start_offset = self.next_prev_offset.map_or(0, |v| v + 1);
+
+            if new_data.offset_interval.start != expected_start_offset {
+                return Err(AppendValidationError::OffsetsAreNotSequential(
+                    OffsetsNotSequentialError::new(
+                        expected_start_offset,
+                        new_data.offset_interval.start,
+                    ),
+                )
+                .into());
+            }
+
+            if new_data.offset_interval.end < new_data.offset_interval.start {
+                invalid_event!(
+                    self.next_data_event.take().unwrap(),
+                    "Invalid offset interval",
+                );
+            }
+        }
+
+        Ok(Decision::Stop)
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Helpers
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct MetadataChainVisitorBatchProcessor {}
+
+impl MetadataChainVisitorBatchProcessor {
+    pub fn get_next_decisions(
+        visitors: BoxedVisitors,
```

Comment:
A bit strange that visitor vec is passed by value in and out

Reply:
Oh, I see now - you are removing visitors that said stop from the vec...

I think the problem is more complex - I think visitors that returned `Decision::NextOfType(X)` should not be called for blocks that are not of type `X` - thus you'll need to store all visitors and decisions until all decisions become `Stop`.

Reply:
WIP

At this point, allocations still remain, I'm in the process of fixing them.

I think I will be able to remove some allocations.

Reply:
All extra allocation have removed -- stack array has been used